subs/pantry/src/Pantry/Storage.hs

```diff
@@ -230,32 +233,76 @@ ModuleName
+  forM_ migrates $ \mig -> logDebug $ "Migration executed: " <> display mig
+  inner $ P.Storage $ withSqliteConnInfo (sqinfo fp False) . runReaderT
   where
-    wrapMigrationFailure = handle (throwIO . MigrationFailure fp)
-    sqinfo fk = set extraPragmas ["PRAGMA busy_timeout=2000;"]
-           $ set fkEnabled fk
+    wrapMigrationFailure fp = handle (throwIO . MigrationFailure fp)
+    sqinfo fp migration
+           = set extraPragmas ["PRAGMA busy_timeout=2000;"]
+
+           -- When doing a migration, we want to disable foreign key
+           -- checking, since the order in which tables are created by
+           -- the migration scripts may not respect foreign keys. The
+           -- rest of the time: enforce those foreign keys.
+           $ set fkEnabled (not migration)
+
+           -- This one is subtle. Enabling Write-Ahead Logging (WAL)
+           -- mode is persistent: the next usage of the database is
+           -- still in WAL mode until explicitly disabled. Therefore,
+           -- when we're already migrated the database, there's no
+           -- need to waste time reenabling WAL. Skipping this also
+           -- allows us to get slightly more meaningful error messages
+           -- if we run into the SQLITE_BUSY bug again.
+           $ set walEnabled migration
+
            $ mkSqliteConnectionInfo (fromString $ toFilePath fp)
 
 withStorage
   :: (HasPantryConfig env, HasLogFunc env)
   => ReaderT SqlBackend (RIO env) a
   -> RIO env a
 withStorage action = do
-  P.Storage pool <- view $ P.pantryConfigL.to P.pcStorage
-  runSqlPool action pool
+  P.Storage run <- view $ P.pantryConfigL.to P.pcStorage
+  run action
+
+-- | Ensure that only one process is trying to write to the database
+-- at a time. See
+-- https://github.com/commercialhaskell/stack/issues/4471 and comments
+-- above.
+withWriteLock :: HasLogFunc env => Path Abs Dir -> RIO env a -> RIO env a
+withWriteLock rootDir inner = do
+  writeLockRelFile <- P.parseRelFile "pantry-write.lock"
+  let lockFile = toFilePath $ rootDir P.</> writeLockRelFile
+  withRunInIO $ \run -> do
+    mres <- withTryFileLock lockFile Exclusive $ const $ run inner
+    case mres of
+      Just res -> pure res
+      Nothing -> do
+        run $ logInfo "Unable to get a write lock on the Pantry database, waiting..."
+        shouldComplainRef <- newIORef True
+        let complainer = fix $ \loop -> do
+              threadDelay $ 60 * 1000 * 1000 -- 1 minute
+              shouldComplain <- readIORef shouldComplainRef
+              when shouldComplain $ do
+                run $ logWarn "Still waiting on the Pantry database write lock..."
+                loop
+            stopComplaining = writeIORef shouldComplainRef False
+            worker = withFileLock lockFile Exclusive $ const $ do
+              run $ logInfo "Acquired the Pantry database write lock"
+              stopComplaining
+              run inner
+        runConcurrently $ Concurrently complainer
```

Comment:
I might be wrong here, but it looks like by using applicative here it can tie up the calling thread unnecessarily for a whole minute. What I mean is whenever `inner` is very quick and finishes early, `complainer` will still be blocked in `threadDelay`. Wouldn't it be better to use Alternative's `<|>` or `race` and make sure `complainer` always calls `loop`, so it gets killed by an exception instead of exiting the loop, i.e.:
```haskell
              when shouldComplain $ do
                run $ logWarn "Still waiting on the Pantry database write lock..."
              loop
```

Reply:
You're right, though I'd prefer not to use a `race` here to avoid async exceptions. Let me take a stab at another approach...